### PR TITLE
Add channels_last for all torchbench models and enable it with run.py

### PR DESCRIPTION
**Pull request by chuanqi129**
In order to standardize the performance evluation and increase coverage, added the `channels_last` for all torchbench models, and enable it with `run.py` entry for debug using. This PR as a first step to standardize and increase coverage for TorchBench, which works for the roadmap #1293.  

Took `alexnet` as an example, which run on CLX 8280L (28cc)
```shell
python run.py alexnet -d cpu -m eager -t eval
Running eval method from alexnet on cpu in eager mode with input batch size 128.
CPU Total Wall Time: 108.189 milliseconds
```

```shell
python run.py alexnet -d cpu -m eager -t eval --channels-last
Running eval method from alexnet on cpu in eager mode with input batch size 128.
CPU Total Wall Time:  72.930 milliseconds
```

### CODE REVIEW

**Comment by yanbing-j** [comment]
Thanks. Generally good to me.
Could you please list the coverage of `channels_last` support in `torchbenchmark` in the description of this PR? And also describe the models that cannot support `channels_last` and why.

**Comment by juliusshufan** [comment]
Hi @xuzhao9,

This PR is the initial implementation reflecting the RFC https://github.com/pytorch/benchmark/issues/1293 for torchbench changes, the changes enable channels_last for all torchbench models w/o model specific scripts change. Would you please comment if we are on the right direction. Thank you for your time.

**Review by xuzhao9 — COMMENTED** [review]
@juliusshufan  It is exciting to see `channels_last` can be supported across all models! It is definitely a step in the right direction.

Can you help address the inline comments?

**Comment by chuanqi129** [comment]
> @juliusshufan It is exciting to see `channels_last` can be supported across all models! It is definitely a step in the right direction.
> 
> Can you help address the inline comments?

Thanks @xuzhao9 for your review comments, and sorry for the delay reply. I will focus on those comments in recent days and get feedback to you ASAP~

File: torchbenchmark/util/model.py [hunk 1]
```diff
@@ -340,3 +340,46 @@ def bench_allclose(a, b):
...
+    def enable_channels_last(self):
+        model_name = inspect.getfile(self.__class__).split(os.sep)[-2]
+        try:
+            model, example_inputs = self.get_module()
+        except NotImplementedError:
+            warnings.warn(UserWarning(f"{model_name} doesn't support `channels_last` yet!"))
+            return
+        try:
+            model = model.to(memory_format=torch.channels_last).eval()
+            if isinstance(example_inputs, torch.Tensor) and example_inputs.dim()==4:
```
[the remark this answers is not in the document]
Comment (chuanqi129) [reply torchbenchmark/util/model.py]: Modified, thanks for the guide~

File: torchbenchmark/util/model.py [hunk 2]
```diff
@@ -340,3 +340,46 @@ def bench_allclose(a, b):
         torch._C._set_graph_executor_optimize(True)
 
         bench_allclose(base, opt)
+
+    def enable_channels_last(self):
+        model_name = inspect.getfile(self.__class__).split(os.sep)[-2]
```
[the remark this answers is not in the document]
Comment (chuanqi129) [reply torchbenchmark/util/model.py]: Modified, thanks

File: torchbenchmark/util/model.py [hunk 3]
```diff
@@ -340,3 +340,46 @@ def bench_allclose(a, b):
...
+        try:
+            model, example_inputs = self.get_module()
+        except NotImplementedError:
+            warnings.warn(UserWarning(f"{model_name} doesn't support `channels_last` yet!"))
+            return
+        try:
+            model = model.to(memory_format=torch.channels_last).eval()
```
[the remark this answers is not in the document]
Comment (chuanqi129) [reply torchbenchmark/util/model.py]: Moved the `.eval()`, also support train mode now

File: torchbenchmark/util/model.py [hunk 4]
```diff
@@ -340,3 +340,46 @@ def bench_allclose(a, b):
...
+                    else:
+                        new_example_inputs[k] = example_inputs[k]
+                example_inputs = new_example_inputs
+        except Exception:
+            pass
+        self.set_module(model)
+        if hasattr(self, 'example_inputs'):
+            self.example_inputs = example_inputs
+        else:
+            warnings.warn(UserWarning(f"{model_name} example inputs doesn't convert to `channels_last`!"))
```
[the remark this answers is not in the document]
Comment (chuanqi129) [reply torchbenchmark/util/model.py]: That's right. But if the model self can convert to channels last but the example_inputs can't convert, we still want to the model can run instead of a error. How do you think?

File: torchbenchmark/util/model.py [hunk 5]
```diff
@@ -340,3 +340,46 @@ def bench_allclose(a, b):
...
+            elif isinstance(example_inputs, dict):
+                new_example_inputs = {}
+                for k in example_inputs.keys():
+                    if isinstance(example_inputs[k], torch.Tensor) and example_inputs[k].dim() ==4:
+                        new_example_inputs[k] = example_inputs[k].to(memory_format=torch.channels_last)
+                    else:
+                        new_example_inputs[k] = example_inputs[k]
+                example_inputs = new_example_inputs
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (chuanqi129) [reply torchbenchmark/util/model.py]: Similar with last comments, add a warning for such case

[the remark this answers is not in the document]
Comment (xuzhao9) [reply torchbenchmark/util/model.py]: makes sense, thanks!

File: torchbenchmark/util/model.py [hunk 6]
```diff
@@ -340,3 +341,25 @@ def bench_allclose(a, b):
...
+        model_name = self.name
+        try:
+            model, _ = self.get_module()
+            model = model.to(memory_format=torch.channels_last)
+        except RuntimeError:
+            warnings.warn(UserWarning(f"{model_name} doesn't support `channels_last` yet!"))
+            return
+        self.set_module(model)
+        def inputs_convert(example_inputs):
+            if isinstance(example_inputs, torch.Tensor) and example_inputs.dim()==4:
```
Comment (xuzhao9) [inline torchbenchmark/util/model.py:355]: Just curious why we are restricting `example_inputs.dim()==4` here?

**Review by xuzhao9 — APPROVED** [review]
LGTM overall, nice enablement!

**Comment by juliusshufan** [comment]
@xuzhao9  Appreciate for your review. May I know your thinkings on next, especially from continous improvement on support more optimization path on CPU and inference? We may then figure out an overall enhancement plan.

**Comment by xuzhao9** [comment]
> @xuzhao9 Appreciate for your review. May I know your thinkings on next, especially from the perspective of continous improvement on support more optimization path on CPU and inference? We may then figure out an overall enhancement plan.

I suggest we can work on adding a TorchBench userbenchmark: https://github.com/pytorch/benchmark/blob/main/userbenchmark/ADDING_USERBENCHMARKS.md.

After adding the userbenchmark, we can enable CI on it, which can automatically generate performance metrics and we can use it to track the continuous improvements.

**Review by davidberard98 — APPROVED** [review]

File: torchbenchmark/util/model.py (hunk 6 above)
Comment (xuzhao9) [reply torchbenchmark/util/model.py:355]: Ah, thanks for pointing it out!

**Comment by chuanqi129** [comment]
Thanks @xuzhao9 and @davidberard98 's review and approval. I saw there is a CI failure in the check `Facebook Internal - Builds & Tests `, but seems I can't check the failure due to the limitation of this check. Could you please help on it? Whether the failure block the PR merge? Thanks again in advance~

**Comment by xuzhao9** [comment]
@chuanqi129 Looks like we have a racing PR trying to land internally, trying to land it again.
